p2p/shim.go

```diff
@@ -0,0 +1,297 @@
+					if src == nil {
+						panic(fmt.Sprintf("failed to proxy envelope; failed to find peer (%s)", e.To))
+					}
+
+					_ = src.Send(cs.Descriptor.ID, bz)
+				}
+			}
+		}(cs)
+	}
+}
+
+// handlePeerErrors iterates over each p2p Channel and starts a separate go-routine
+// where we listen for peer errors. For each peer error, we find the peer from
+// the legacy p2p Switch and execute a StopPeerForError call with the corresponding
+// peer error.
+func (rs *ReactorShim) handlePeerErrors() {
+	if rs.Switch == nil {
+		panic("handlePeerErrors: reactor shim switch is nil")
+	}
+
+	for _, cs := range rs.Channels {
+		go func(cs *ChannelShim) {
+			for pErr := range cs.PeerErrCh {
+				peer := rs.Switch.peers.Get(ID(pErr.PeerID.String()))
+				if peer == nil {
+					panic(fmt.Sprintf("failed to handle peer error; failed to find peer (%s)", pErr.PeerID))
+				}
+
+				rs.Switch.StopPeerForError(peer, pErr.Err)
+			}
+		}(cs)
+	}
+}
+
+// OnStart executes the reactor shim's OnStart hook where we start all the
+// necessary go-routines in order to proxy peer envelopes and errors per p2p
+// Channel.
+func (rs *ReactorShim) OnStart() error {
+	rs.proxyPeerEnvelopes()
+	rs.handlePeerErrors()
+	return nil
+}
+
+// OnStop executes the reactor shim's OnStop hook where all inbound p2p Channels
+// are closed and the PeerUpdateCh is closed. The caller must be sure to also
+// stop the real reactor so the shim's go routines can exit successfully.
+func (rs *ReactorShim) OnStop() {
+	for _, cs := range rs.Channels {
+		close(cs.InCh)
+	}
+
+	close(rs.PeerUpdateCh)
+}
+
+// GetChannel returns a p2p Channel reference for a given ChannelID. If no
+// Channel exists, nil is returned.
+func (rs *ReactorShim) GetChannel(cID ChannelID) *Channel {
+	channelShim, ok := rs.Channels[cID]
+	if ok {
+		return channelShim.Channel
+	}
+
+	return nil
+}
+
+// GetChannels implements the legacy Reactor interface for getting a slice of all
+// the supported ChannelDescriptors.
+func (rs *ReactorShim) GetChannels() []*ChannelDescriptor {
+	descriptors := make([]*ChannelDescriptor, len(rs.Channels))
+
+	sortedChIDs := make([]ChannelID, 0, len(descriptors))
+	for cID := range rs.Channels {
+		sortedChIDs = append(sortedChIDs, cID)
+	}
+
+	sort.Slice(sortedChIDs, func(i, j int) bool { return sortedChIDs[i] < sortedChIDs[j] })
+
+	for i, cID := range sortedChIDs {
+		descriptors[i] = rs.Channels[cID].Descriptor
+	}
+
+	return descriptors
+}
+
+// AddPeer sends a PeerUpdate with status PeerStatusUp on the PeerUpdateCh.
+// The embedding reactor must be sure to listen for messages on this channel to
+// handle adding a peer.
+func (rs *ReactorShim) AddPeer(peer Peer) {
+	peerID, err := PeerIDFromString(string(peer.ID()))
+	if err != nil {
+		// It is OK to panic here as we'll be removing the Reactor interface and
+		// Peer type in favor of using a PeerID directly.
+		panic(err)
+	}
+
+	go func() {
+		rs.PeerUpdateCh <- PeerUpdate{PeerID: peerID, Status: PeerStatusUp}
```

Comment:
will this panic once PeerUpdateCh is closed?

Reply:
Yes, but the `PeerUpdateCh` is closed only once the reactor shim is stopped `OnStop`, so unless we call `AddPeer|RemovePeer` _after_ the reactor is stopped, this shouldn't be a problem.

Why would these pile up @erikgrinaker? The real reactor process messages from this channel and handles them accordingly. They would only pile up if the real reactor isn't listening for these messages.

Would you still like this to block?

Reply:
> Yes, but the PeerUpdateCh is closed only once the reactor shim is stopped (via OnStop), so unless we call AddPeer|RemovePeer after the reactor is stopped, this shouldn't be a problem.

what if there are too many peer updates and this one blocks? then something happens and reactor stops, the goroutine will panic, which will lead to the whole node panicking.

Reply:
Yeah, I think your reasoning is correct. If there is a lot of back pressure of peer updates and as a result we spawn a significant number of go routines that send on `PeerUpdateCh`, a bunch of them could block since there is a single reader and _iff_ that reader closes/stops and there are still lingering go routines sending on `PeerUpdateCh`, yes I believe this would yield in panic(s).

I guess we'll block here...